### PR TITLE
refactor(NODE-5352): refactor AbstractOperation to use async

### PR DESCRIPTION
**Pull request by malikj2000**
#### What is changing?
created callback
subclass of operation
and changed execute
##### Is there new documentation needed for these changes?
no
#### What is the motivation for this change?
first step in asyncifying the operations layer

### Double check the following

- [ ] Ran `npm run check:lint` script
- [ ] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [ ] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [ ] Changes are covered by tests
- [ ] New TODOs have a related JIRA ticket

### CODE REVIEW

File: src/operations/operation.ts
Comment (baileympearson) [inline src/operations/operation.ts:121]:
```suggestion
export abstract class AbstractCallbackOperation<TResult> extends AbstractOperation<TResult> {
```

**Review by baileympearson — CHANGES_REQUESTED** [review]
minor comments

File: src/operations/operation.ts
Comment (W-A-James) [reply src/operations/operation.ts:109]: Once you get it to build, you're probably going to see some test failures that I flagged below this.

**Review by W-A-James — CHANGES_REQUESTED** [review]

File: src/operations/operation.ts
Comment (W-A-James) [reply src/operations/operation.ts]: Why remove the default generic parameter?

Comment (baileympearson) [reply src/operations/operation.ts]: Because it's already defaulted in `AbstractOperation`.

Comment (W-A-James) [reply src/operations/operation.ts]: In that case I think we should keep it here and remove it from `AbstractOperation` because TypeScript doesn't understand that we have it defaulted for the subclasses of `AbstractCallbackOperation` that don't explicitly provide the parameter when they are declared so we're getting TypeScript errors.

File: src/operations/execute_operation.ts
Comment (malikj2000) [reply src/operations/execute_operation.ts:28]: Yeah we missed one, I changed it

File: src/operations/execute_operation.ts
Comment (W-A-James) [reply src/operations/execute_operation.ts:34]: @malikj2000 my mistake on asking you to change that then.

**Review by W-A-James — CHANGES_REQUESTED** [review]
@malikj2000 There are test failures to address

File: test/unit/cmap/connection.test.ts
Comment (W-A-James) [inline test/unit/cmap/connection.test.ts:76]: For my own understanding, why was this changed?

File: test/unit/sdam/server.test.ts
Comment (W-A-James) [inline test/unit/sdam/server.test.ts:84]: This test is failing.

**Review by W-A-James — CHANGES_REQUESTED** [review]
In addition to the comments above, there are also some failing unit tests for operations that subclass `AbstractCallbackOperation`

File: test/unit/cmap/connection.test.ts
Comment (baileympearson) [reply test/unit/cmap/connection.test.ts:76]: Malik and I changed this debugging, but now this is unnecessary.  We should revert any changes that aren't necessary for this PR

File: test/unit/operations/find.test.ts
Comment (W-A-James) [inline test/unit/operations/find.test.ts:46]: You can get rid of the `promisify` here instead of replacing `execute` with `executeCallback`. Remember that what `promisify` does is take an asynchronous function that returns its results via callback and changes it into one that returns a promise instead. That's exactly what you did with `execute` itself, so you can just await that directly.

The same goes for all the other unit tests like this below.

**Review by W-A-James — CHANGES_REQUESTED** [review]

**Review by W-A-James — DISMISSED** [review]

**Review by baileympearson — CHANGES_REQUESTED** [review]

**Review by baileympearson — APPROVED** [review]
waiting on CI after the rebase before merging but otherwise LGTM

**Pull request merged**
